File: build-integration/elixir-mix/sample-consumer/lib/robot_demo.ex
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 ZeroDDS Contributors
#
# Round-trips a Robot.Pose through the *generated* marshal_xcdr/2 and
# unmarshal/2 — proof that the :zerodds_idl compiler actually ran before
# :elixir compiled this module's own dependency on Robot.Pose.
defmodule RobotDemo do
  def main do
    pose = %Robot.Pose{robot_id: "r2d2", x: 1.5, y: -2.5, theta: 0.75}

    wire = Robot.Pose.marshal_xcdr(pose, :little)
    decoded = Robot.Pose.unmarshal(wire, :little)

    if decoded.robot_id != pose.robot_id or decoded.x != pose.x or decoded.y != pose.y or
         decoded.theta != pose.theta do
      raise "round-trip mismatch: #{inspect(decoded)}"
    end

    IO.puts(
      "OK: Robot.Pose round-tripped through generated marshal_xcdr/unmarshal (#{byte_size(wire)} wire bytes), robot_id=#{decoded.robot_id}"
    )
  end
end

File: build-integration/julia-build/sample-consumer/src/RobotDemo.jl
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 ZeroDDS Contributors
#
# Round-trips a Pose through the *generated* marshal_xcdr/unmarshal_xcdr_Pose
# — proof that deps/build.jl actually ran zerodds-idlc before this module
# was loaded (src/gen/Robot.jl does not exist in the source tree).
module RobotDemo

include(joinpath(@__DIR__, "gen", "Robot.jl"))

function main()
    pose = Pose("r2d2", 1.5, -2.5, 0.75)

    wire = marshal_xcdr(pose, LE)
    decoded = unmarshal_xcdr_Pose(wire, LE)

    @assert decoded.robot_id == pose.robot_id
    @assert decoded.x == pose.x
    @assert decoded.y == pose.y
    @assert decoded.theta == pose.theta

    println("OK: Pose round-tripped through generated marshal_xcdr/unmarshal_xcdr_Pose (",
            length(wire), " wire bytes), robot_id=", decoded.robot_id)
end

end # module RobotDemo

RobotDemo.main()
